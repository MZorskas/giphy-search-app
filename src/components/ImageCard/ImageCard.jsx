import React, { useState, useRef, memo } from 'react';
import './ImageCard.scss';
import ImageModal from './ImageModal';
import PropTypes from 'prop-types';

const ImageCard = memo(({ images, alt, elementRef }) => {
  const [isOpen, setIsOpen] = useState(false);

  const imageCardRef = useRef(null);

  const closeModal = () => {
    document.body.removeAttribute('style');
    setIsOpen(false);
    imageCardRef.current.focus();
  };

  const handleMouseDown = (event) => event.preventDefault();

  const handleClick = () => {
    if (!isOpen) {
      document.body.style.overflow = 'hidden';
      setIsOpen(true);
    }
  };

  const handleKeyDown = (event) => {
    if (event.key === 'Enter') setIsOpen(true);
  };

  return (
    <div
      className="image-card"
      tabIndex="0"
      onClick={handleClick}
      onKeyDown={handleKeyDown}
      ref={imageCardRef}
      data-testid="image-card"
      onMouseDown={handleMouseDown}
    >
      <picture>
        <source
          media="(max-width: 768px)"
          srcSet={images?.downsized.url}
          data-testid="image-card-source-1"
        />
        <source
          media="(min-width: 769px)"
          srcSet={images?.downsized_large.url}
          data-testid="image-card-source-2"
        />
        <img
          src={images?.downsized_large.url}
          alt={alt}
          ref={elementRef}
          className="image-card__image"
        />
      </picture>

      {isOpen && (
        <ImageModal
          imageUrl={images?.original.url}
          alt={alt}
          closeModal={closeModal}
        />
      )}
    </div>
  );
});

ImageCard.propTypes = {
  alt: PropTypes.string,
  images: PropTypes.object.isRequired,
  elementRef: PropTypes.func,
};

export default ImageCard;
