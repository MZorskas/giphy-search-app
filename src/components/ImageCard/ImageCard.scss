@import 'src/styles/mixins/index';
@import 'src/styles/base/base';

.image-card {
  position: relative;
  transition: 0.5s;

  &__image {
    @include position-absolute(0, null, 0, 0);

    margin: auto;
    max-height: 100%;
    max-width: 100%;
    width: 100%;
  }

  &::before {
    background-color: $color-black;
    content: '';
    display: block;
    padding-top: 100%;
    opacity: 0.4;
  }

  &:hover {
    box-shadow: 0 0.3rem 0.2rem rgba(52, 232, 158, 0.034),
      0 0.7rem 0.5rem rgba(52, 232, 158, 0.048),
      0 1.3rem 1rem rgba(52, 232, 158, 0.06),
      0 2.2rem 1.8rem rgba(52, 232, 158, 0.072),
      0 4.2rem 3.3rem rgba(52, 232, 158, 0.086),
      0 10rem 8rem rgba(52, 232, 158, 0.12);
  }
}
