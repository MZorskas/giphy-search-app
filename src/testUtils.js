export const generateModalRoot = () => {
  const modalRoot = global.document.createElement('div');
  modalRoot.setAttribute('id', 'modal-root');
  const body = global.document.querySelector('body');
  body.appendChild(modalRoot);
};
