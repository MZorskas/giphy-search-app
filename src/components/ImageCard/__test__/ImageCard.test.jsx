import React from 'react';
import ReactDom from 'react-dom';
import ImageCard from '../ImageCard';
import { generateModalRoot } from 'testUtils';
import { render, fireEvent, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';

const images = {
  downsized: {
    url: 'imgURL1',
  },
  downsized_large: {
    url: 'imgURL2',
  },
  original: {
    url: 'imgURL3',
  },
};

const alt = 'image alt';

describe('ImageCard component', () => {
  it('renders with propTypes warning', () => {
    const spy = jest.spyOn(global.console, 'error');
    const div = document.createElement('div');
    ReactDom.render(<ImageCard />, div);
    expect(spy).toHaveBeenCalled();
  });

  it('renders without crashing', () => {
    const spy = jest.spyOn(global.console, 'error');
    const div = document.createElement('div');
    ReactDom.render(<ImageCard images={images} />, div);
    expect(spy).not.toHaveBeenCalled();
  });

  it('renders with correct alt attribute', () => {
    render(<ImageCard images={images} alt={alt} />);
    expect(screen.getByAltText(alt)).toBeInTheDocument();
  });

  it('renders with correct srcSet', () => {
    render(<ImageCard images={images} alt={alt} />);

    expect(
      screen.getByTestId('image-card-source-1').getAttribute('srcSet')
    ).toMatch(images.downsized.url);
    expect(
      screen.getByTestId('image-card-source-2').getAttribute('srcSet')
    ).toMatch(images.downsized_large.url);

    expect(screen.getByAltText(alt).getAttribute('src')).toMatch(
      images.downsized_large.url
    );
  });

  it('renders modal on mouse click', () => {
    generateModalRoot();
    render(<ImageCard images={images} />);
    expect(screen.queryByTestId('image-modal')).not.toBeInTheDocument();

    fireEvent.click(screen.getByTestId('image-card'));

    expect(screen.queryByTestId('image-modal')).toBeInTheDocument();
  });

  it('renders modal on enter key down', () => {
    generateModalRoot();
    render(<ImageCard images={images} />);
    expect(screen.queryByTestId('image-modal')).not.toBeInTheDocument();

    fireEvent.keyDown(screen.getByTestId('image-card'), { key: 'Enter' });

    expect(screen.queryByTestId('image-modal')).toBeInTheDocument();
  });

  it('matches snapshot', () => {
    const tree = renderer
      .create(<ImageCard alt={alt} images={images} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
