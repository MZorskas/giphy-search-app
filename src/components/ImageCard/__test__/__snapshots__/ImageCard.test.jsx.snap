// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageCard component matches snapshot 1`] = `
<div
  className="image-card"
  data-testid="image-card"
  onClick={[Function]}
  onKeyDown={[Function]}
  onMouseDown={[Function]}
  tabIndex="0"
>
  <picture>
    <source
      data-testid="image-card-source-1"
      media="(max-width: 768px)"
      srcSet="imgURL1"
    />
    <source
      data-testid="image-card-source-2"
      media="(min-width: 769px)"
      srcSet="imgURL2"
    />
    <img
      alt="image alt"
      className="image-card__image"
      src="imgURL2"
    />
  </picture>
</div>
`;
